### ISSUE TITLE
Custom migrations source (for bundling)

### ISSUE THREAD

**Issue by IlyaSemenov** [romeerez/orchid-orm#155]
I would like to be able to run rake-db migrations in production. In production, I don't distribute Typescript sources, I use 2-phase dockerfiles and the running container only has bundles compiled with esbuild/rollup.

However, the only way rake-db discovers migrations is `readdir`:
https://github.com/romeerez/orchid-orm/blob/4c436cf35d6a0cdd7baf6abfd85bdc179aa959fb/packages/rake-db/src/common.ts#L315-L323

I would appreciate if there was a way to provide the set of migrations programmatically, similar to how Knex does that: https://knexjs.org/guide/migrations.html#custom-migration-sources

Basically, if the above code could call a custom callback instead of always doing `await readdir(migrationsPath)` that would probably suffice (then the results could be imported with the custom `import` callback).

**Comment by romeerez** [comment]
Makes sense, rake-db should be capable of running compiled js files, which is not currently supported.

But why provide anything programmatically?
If `rake-db` will use `readdir` to find out compiled migrations and run them, would you need something else?

**Comment by IlyaSemenov** [comment]
Because I [bundle](https://esbuild.github.io/getting-started/#your-first-bundle) sources. 

For each executable script that is needed in production (a server or a cli script) this generates exactly one file such as `dist/dbScript.js` which combines all project files (and sometimes all node_module sources, too, although at the moment I'm not doing that).

The sources are coming from a mono-repo with dozens of packages, and each micro-service only needs a subset of them, so bundling is very practical to minimize footprint. This also allows simpler serverless deploy, and reduces startup times.

**Comment by romeerez** [comment]
Understood, but still, it's unclear how passing an array of migrations is going to help.

If we imagine that `rake-db` has exactly the same `MigrationSource` interface:

```ts
class MyMigrationSource {
  getMigrations() {
    // resolves array of migration file names
    return Promise.resolve(['migration1'])
  }

  getMigrationName(migration) {
    // get the version from a file name
    return migration;
  }

  getMigration(migration) {
    switch(migration) {
      case 'migration1':
        // get functions to migrate up or down by a migration name
        return {
          up(knex)   { /* ... */ },
          down(knex) { /* ... */ },
        }
    }
  }
}
```

Would this be enough? Is this similar to how you did this with Knex?

I'm hoping to find a user-friendly way for this.

Thinking, what if `readdir` will be used in esbuild config to compile all migration files separately, and `rake-db` will work with js files. It could be wrapped into a plugin for esbuild.

Or, `rake-db` could have array of lazy imports of migration files, and the `db new` command will also automatically insert a new migration path to this array.

**Comment by IlyaSemenov** [comment]
Sure, that would be flexible enough to accommodate to any needs. Perhaps too detailed for this particular use case, as I didn't want to fiddle with migration names — the current logic of parsing filenames suits just fine.

This is how I could load migrations so that it works in both dev environment and in the bundle:

```ts
const migrations = import.meta.glob('./migrations/*.ts')
```

This returns an object shaped like:

```
{
  './migrations/20230811075433_init.ts': [Function: ./migrations/20230811075433_init.ts]
}
```

...where the function, when called, returns a promise that imports the respective module.

So basically what would be enough for my case is:

```ts
const migrations = import.meta.glob('./migrations/*.ts')

export const change = rakeDb(
  { databaseURL: db_url },
  {
    baseTable: BaseTable,
    migrationsSource: Object.keys(migrations),
    import: path => migrations[path],
  },
)
```

**Comment by IlyaSemenov** [comment]
Actually, I don't anticipate any other combination of `migrations` and `import`, so perhaps `rakeDb` could simply accept an object of `{ [migrationName: string]: () => Promise<Module> }`:

```ts
export const change = rakeDb(
  { databaseURL },
  {
    baseTable: BaseTable,
    migrations: import.meta.glob('./migrations/*.ts'),
  },
)
```

**Comment by romeerez** [comment]
Ready! See [notes and example here](https://github.com/romeerez/orchid-orm-examples/tree/main/packages/express-esm).

You need to add `migrations` just as you suggested, thanks for letting me know about `import.meta.glob`:
```ts
migrations: import.meta.glob('./migrations/*.ts'),
```

`migrationsPath` option is still used to generate new migrations into.

**Comment by IlyaSemenov** [comment]
Thanks! I'll try to integrate this in a couple days.
